src/rust/engine/async_semaphore/src/lib.rs

```diff
@@ -50,10 +59,16 @@ impl AsyncSemaphore {
       inner: Arc::new(Mutex::new(Inner {
         waiters: VecDeque::new(),
         available_permits: permits,
+        task_id: 0,
       })),
     }
   }
 
+  pub fn num_waiters(&self) -> usize {
```

Comment:
Add a `#[cfg(test)]` to this function?

Production code should always have the same lock acquired for both finding out how many waiters there are, and making decisions based on that number, so this is only really a useful function in tests (which is also the only place we use it :)).

Reply:
This one is useful outside of tests for the purposes of UI/UX: the log message that @hrfuller added here should probably move out to our consumer.